codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/JsonShapeDeserVisitor.java

```diff
@@ -0,0 +1,245 @@
+        writer.openBlock("for decoder.More() {", "}", () -> {
+            // Deserialize the key
+            writer.write("token, err := decoder.Token()");
+            writer.write("if err != nil { return err }");
+            writer.write("");
+            writer.write("key, ok := token.(string)");
+            writer.write("if !ok { return fmt.Errorf(\"expected map-key of type string, found type %T\", token)}");
+            writer.write("");
+
+            // Deserialize the value. We need to write out an intermediate variable here
+            // since we can't just pass in &mv[key]
+            writer.write("var parsedVal $P", targetSymbol);
+            context.getModel().expectShape(member.getTarget()).accept(getMemberDeserVisitor(member, "parsedVal"));
+            writer.write("mv[key] = parsedVal");
+            writer.write("");
+        });
+
+        writeJsonTokenizerEndStub(writer, shape);
+        writer.write("*v = mv");
+        writer.write("return nil");
+    }
+
+    @Override
+    protected void deserializeStructure(GenerationContext context, StructureShape shape) {
+        GoWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Symbol symbol = symbolProvider.toSymbol(shape);
+
+        writeJsonTokenizerStartStub(writer, shape);
+
+        // Initialize the value now that the start stub has verified that there's something there.
+        writer.write("var sv $P", symbol);
+        writer.openBlock("if *v == nil {", "", () -> {
+            writer.write("sv = &$T{}", symbol);
+            writer.openBlock("} else {", "}", () -> {
+                writer.write("sv = *v");
+            });
+        });
+
+        // Iterate through the decoder. The member visitor will handle popping tokens.
+        writer.openBlock("for decoder.More() {", "}", () -> {
+            writer.write("t, err := decoder.Token()");
+            writer.write("if err != nil { return err }");
+            writer.openBlock("switch t {", "}", () -> {
+                Set<MemberShape> members = new TreeSet<>(shape.members());
+                for (MemberShape member : members) {
+                    String memberName = symbolProvider.toMemberName(member);
+                    String serializedMemberName = getSerializedMemberName(member);
+                    writer.openBlock("case $S:", "", serializedMemberName, () -> {
+                        String dest = "sv." + memberName;
+                        context.getModel().expectShape(member.getTarget()).accept(getMemberDeserVisitor(member, dest));
+                    });
+                }
+
+                writer.openBlock("default:", "", () -> {
+                    // TODO: do I need to import from restjson here?
+                    writer.addUseImports(AwsGoDependency.AWS_REST_JSON_PROTOCOL);
+                    writer.write("err := restjson.DiscardUnknownField(decoder)");
```

Comment:
Can probably move the `DiscardUnknownField` util to smithy-go/Json package

Reply:
Yeah I have a feeling most of the rest json stuff can be moved up. I end up importing it in several other places too.